File: typescript/packages/core/src/observers.ts
/** Model observers. */

import type { Device } from "./models/device";

/** Lifecycle hooks for {@link Device}; the app registers one implementor. */
export interface DeviceObserver {
    created(device: Device): void;
    updated?(device: Device): void;
    deleted?(device: Device): void;
}

/** Writes an audit line whenever a device appears. */
export class AuditingDeviceObserver implements DeviceObserver {
    readonly lines: string[] = [];

    created(device: Device): void {
        this.lines.push(`device.created ${device.label()}`);
    }

    deleted(device: Device): void {
        this.lines.push(`device.deleted ${device.label()}`);
    }
}

File: rust/crates/atelier-core/src/observers.rs
//! Model observers.

use crate::models::device::Device;

/// Lifecycle hooks for [`Device`]; the app registers one implementor.
pub trait DeviceObserver: Send + Sync {
    fn created(&self, device: &Device);

    fn updated(&self, _device: &Device) {}

    fn deleted(&self, _device: &Device) {}
}

/// Writes an audit line whenever a device appears.
#[derive(Debug, Default)]
pub struct AuditingDeviceObserver {
    pub lines: std::sync::Mutex<Vec<String>>,
}

impl DeviceObserver for AuditingDeviceObserver {
    fn created(&self, device: &Device) {
        if let Ok(mut lines) = self.lines.lock() {
            lines.push(format!("device.created {}", device.label()));
        }
    }

    fn deleted(&self, device: &Device) {
        if let Ok(mut lines) = self.lines.lock() {
            lines.push(format!("device.deleted {}", device.label()));
        }
    }
}
